src/dbami/helpers/apply_migrations.py

```diff
@@ -0,0 +1,277 @@
+            The migration version to target. If unset, the highest-known migration
+            version will be targeted.
+
+        do_rollback (bool):
+            Whether to roll back to an older migration version. Must be `True` if the
+            target migration version is older than the current version.
+
+        use_migration_lock (bool):
+            Whether to use an advisory lock to prevent concurrent migrations.
+
+        migration_lock_acquisition_timeout_ms (int):
+            Timeout (in milliseconds) to wait for an advisory lock to be successfully
+            acquired. A value of 0 disables the timeout, meaning the wait for a lock
+            will continue indefinitely. Has no effect unless `use_migration_lock` is
+            `True`.
+
+        revoke_connect_on_role_names (Sequence[str]):
+            List of role names to revoke `CONNECT` privileges on before running
+            migrations. This prevents other users/services from reconnecting during
+            migrations.
+
+        connection_wait_max_polls (Optional[int]):
+            Maximum number of times to check (poll) if revoked role connections are
+            closed before aborting or attempting to force-close the connections. Set to
+            `None` to wait (poll) indefinitely or <= 0 to disable waiting (polling).
+
+        connection_wait_poll_interval_ms (int):
+            Polling interval (in milliseconds) for checking if revoked role connections
+            are closed. Values less than 100 will be coerced to 100. Has no effect if
+            `connection_wait_max_polls` is set to <= 0.
+
+        force_close_connections_timeout_ms (Optional[int]):
+            Timeout (in milliseconds) to wait for force-closed connections to
+            successfully close. Set to `None` to disable force-closing or <= 0 to
+            proceed without waiting.
+    """
+
+    target_migration_version: Optional[int] = None
+    do_rollback: bool = False
+    use_migration_lock: bool = True
+    migration_lock_acquisition_timeout_ms: int = 10000
+    revoke_connect_on_role_names: Sequence[str] = field(default_factory=list)
+    connection_wait_max_polls: Optional[int] = None
+    connection_wait_poll_interval_ms: int = 2000
+    force_close_connections_timeout_ms: Optional[int] = 10000
+
+    def __post_init__(self) -> None:
+        if self.connection_wait_poll_interval_ms < 100:
+            self.connection_wait_poll_interval_ms = 100
+
+    @property
+    def wait_for_other_connections_to_close(self) -> bool:
+        return (
+            self.connection_wait_max_polls is None or self.connection_wait_max_polls > 0
+        )
+
+
+class MigrationHelper(Helper[MigrationHelperConfig]):
+    @classmethod
+    def get_config_class(cls) -> Type[MigrationHelperConfig]:
+        return MigrationHelperConfig
+
+    async def wait_for_other_connections_to_close(
+        self,
+        conn: asyncpg.Connection,
+        poll_interval_ms: int = 2000,
```

Comment:
suggestion: consider making the poll interval seconds, since this should never be set to less than 1s

Reply:
>should never be set to less than 1s

Note: I think it is never less than 0.1s (100 milliseconds)

I'm going to defer to @jkeifer on this. We kept all time units in these backend classes in milliseconds for consistency.

Reply:
I don't totally remember my rationale for using ms. I think it was motivated by two things: 1) some of the places we are using time I thought it might be good to support sub-second durations, at least initially, and 2) because the backing pg functions for some of the operations also take durations in ms.

That said, I don't feel strongly about which units to use. If we want to switch the config to seconds, then we can do so. The thing I do feel strongly about: I want to see the time units for all duration parameters be consistent (unless we have a really good reason not be consistent somewhere), so that would mean changing the units on all those duration parameters if we wanted to use seconds.

Reply:
I vote to keep it as is: seconds for the CLI interface (user friendly), milliseconds in the migration helper (match what postgres uses).